### PR TITLE
convert SHOW MEASUREMENTS to a distributed query

### PR DESCRIPTION
**Pull request by dgnorton**

### CODE REVIEW

**Comment by otoolep** [comment]
Cool, let me know when you want review feedback.

File: cluster/service.go [hunk 1]
```diff
@@ -37,7 +38,7 @@ type Service struct {
 	TSDBStore interface {
 		CreateShard(database, policy string, shardID uint64) error
 		WriteToShard(shardID uint64, points []tsdb.Point) error
-		CreateMapper(shardID uint64, query string, chunkSize int) (tsdb.Mapper, error)
+		CreateMapper(shardID uint64, stmt influxql.Statement, chunkSize int) (tsdb.Mapper, error)
```
Comment (otoolep) [inline cluster/service.go]: Why this change?

File: meta/data.go [hunk 2]
```diff
@@ -662,6 +662,26 @@ func (di DatabaseInfo) RetentionPolicy(name string) *RetentionPolicyInfo {
 	return nil
 }
 
+// ShardInfos returns a list of all shards' info for the database.
+func (di DatabaseInfo) ShardInfos() []ShardInfo {
```
Comment (otoolep) [inline meta/data.go]: This looks very like what @benbjohnson coded for `SHOW SHARD`. You will want to ignore shards in expired shard groups.

File: tsdb/query_executor.go [hunk 3]
```diff
@@ -547,63 +550,63 @@ func (q *QueryExecutor) filterShowSeriesResult(limit, offset int, rows influxql.
...
 
-func (q *QueryExecutor) executeShowMeasurementsStatement(stmt *influxql.ShowMeasurementsStatement, database string) *influxql.Result {
-	// Find the database.
-	db := q.Store.DatabaseIndex(database)
-	if db == nil {
-		return &influxql.Result{}
+// PlanShowMeasurements creates an execution plan for the given SelectStatement and returns an Executor.
+func (q *QueryExecutor) PlanShowMeasurements(stmt *influxql.ShowMeasurementsStatement, database string, chunkSize int) (Executor, error) {
+	// Get the database info.
+	fmt.Printf("database = %s\n", database)
```
Comment (otoolep) [inline tsdb/query_executor.go]: Obviously this should be removed before commit.

File: tsdb/query_executor.go [hunk 4]
```diff
@@ -547,63 +550,63 @@ func (q *QueryExecutor) filterShowSeriesResult(limit, offset int, rows influxql.
...
-	sort.Sort(measurements)
 
-	offset := stmt.Offset
-	limit := stmt.Limit
-
-	// If OFFSET is past the end of the array, return empty results.
-	if offset > len(measurements)-1 {
-		return &influxql.Result{}
-	}
+	executor := NewShowMeasurementsExecutor(stmt, mappers, chunkSize)
```
Comment (otoolep) [inline tsdb/query_executor.go]: Are you planning a different executor for each statement? Will that be necessary?

**Comment by otoolep** [comment]
Are we going to have new mappers for every meta-query?

File: cluster/service.go (hunk 1 above)
Comment (dgnorton) [inline cluster/service.go]: To avoid unnecessary reparsing of the statement.

File: tsdb/query_executor.go (hunk 4 above)
Comment (dgnorton) [inline tsdb/query_executor.go]: For now, yes.  Maybe once it's all written it will be obvious where some duplicate code can be removed.  I started down the path of trying to avoid this and it became messy.  Also, doing it this way is less risky because no big changes have to be made to `SelectMapper` (used to be `LocalMapper`) or that part of the query engine.

**Comment by dgnorton** [comment]
@otoolep  yes, for now we will have executors and mappers for each statement.  Later we can probably refactor / combine some of the common stuff.

File: tsdb/query_executor.go (hunk 3 above)
Comment (dgnorton) [inline tsdb/query_executor.go]: Good catch.

File: meta/data.go (hunk 2 above)
Comment (dgnorton) [inline meta/data.go]: Good point.  I'll look at Ben's.

Comment (dgnorton) [inline meta/data.go]: Both are doing the same thing.  If his merges first, then I'll update his to call this func or if this merges first, he can do the update.

**Comment by dgnorton** [comment]
@otoolep ready for final review

**Comment by otoolep** [comment]
Can we add more testing? There is no testing of queries across stores, which might help you be sure the code is working as you expected.

What would be great is a test like this one:

https://github.com/influxdb/influxdb/blob/master/tsdb/executor_test.go#L253

which would create two completely distinct stores, write points to them, and then execute `SHOW MEASUREMENTS` across those stores. This is very close to distributed queries -- everything except the network portion. As you add more metaquery functionality, we can add more queries to this test.

The reason you must create two stores is that creating two shards on a single node results only in a single store. So two stores must be created.

**Comment by dgnorton** [comment]
@otoolep added a couple tests.

File: tsdb/executor_test.go [hunk 5]
```diff
@@ -432,6 +432,82 @@ func TestWritePointsAndExecuteTwoShardsTagSetOrdering(t *testing.T) {
...
+		stmt      string // Query statement
+		chunkSize int    // Chunk size for driving the executor
+		expected  string // Expected results, rendered as a string
+	}{
+		{
+			stmt:     `SHOW MEASUREMENTS`,
+			expected: `[{"name":"measurements","columns":["name"],"values":[["cpu"],["mem"]]}]`,
+		},
+		{
+			stmt:     `SHOW MEASUREMENTS WHERE host='serverB'`,
```
Comment (otoolep) [inline tsdb/executor_test.go]: Let's add 1 more test `SHOW MEASUREMENTS WHERE host='serverX'.

Comment (dgnorton) [inline tsdb/executor_test.go]: To show that no measurements are returned?

**Comment by otoolep** [comment]
+1 once the final test been added -- thanks.

*Hopefully* as we add more meta queries, significant refactoring will be possible. A new mapper for each command (it may be the only way, I don't know) would mean a lot of code. I do hope a pattern emerges that means so much code won't be written.

File: tsdb/executor_test.go (hunk 5 above)
Comment (otoolep) [inline tsdb/executor_test.go]: Yeah, just to show nothing blows up when no measurements are returned from anywhere in the cluster. Unless you have that covered.
